step-framework-parent/step-framework-collections/src/main/java/step/core/accessors/AbstractAccessor.java

```diff
@@ -131,12 +143,80 @@ private Equals byId(ObjectId id) {
 
 	@Override
 	public T save(T entity) {
+		if (versionedCollectionDriver != null) {
+			versionedCollectionDriver.save(getVersionForEntity(entity));
+		}
 		return collectionDriver.save(entity);
 	}
 
 	@Override
 	public void save(Iterable<T> entities) {
-		collectionDriver.save((Iterable<T>)entities);
+		if (versionedCollectionDriver != null) {
+			List<VersionableEntity> versionedEntities = new ArrayList<>();
+			entities.forEach(e->versionedEntities.add(getVersionForEntity(e)));
+			versionedCollectionDriver.save(versionedEntities);
+		}
+		collectionDriver.save(entities);
+	}
+
+	private VersionableEntity getVersionForEntity(T entity) {
+		Long current = System.currentTimeMillis();
+		Long currentTsGroup = ((long) (current / newVersionThresholdMs)) * newVersionThresholdMs;
+		//If current version of entity is older than 1 minute create a new one
+		String versionId = entity.getCustomField(VERSION_CUSTOM_FIELD, String.class);
+		VersionableEntity versionableEntity = (versionId != null && ((versionableEntity = getVersionById(versionId)) != null)
+				&& versionableEntity.getUpdateGroupTime() == currentTsGroup) ?
+				versionableEntity : new VersionableEntity();
+
+		versionableEntity.setUpdateTime(current);
+		versionableEntity.setUpdateGroupTime(currentTsGroup);
+		versionableEntity.setEntity(entity);
+		entity.addCustomField(VERSION_CUSTOM_FIELD, versionableEntity.getId().toHexString());
+		return versionableEntity;
+	}
+
+	private VersionableEntity getVersionById(String versionId) {
+		return versionedCollectionDriver.find(byId(new ObjectId(versionId)), null, null, 1, 0).findFirst().orElse(null);
+	}
+
+	@Override
+	public Stream<VersionableEntity> getHistory(ObjectId id, Integer skip, Integer limit) {
+		if (versionedCollectionDriver != null) {
+			SearchOrder order = new SearchOrder(AbstractIdentifiableObject.ID, -1);
+			return versionedCollectionDriver.find(getHistoryFilterById(id), order, skip, limit, 0);
+		} else {
+			throw getVersioningNotSupportedException();
+		}
+	}
+
+	@Override
+	public T restoreVersion(ObjectId entityId, ObjectId versionId) {
+		if (versionedCollectionDriver != null) {
+			VersionableEntity versionableEntity = versionedCollectionDriver.find(byId(versionId),
+					null, null, null, 0).findFirst().orElse(null);
+			T entity = (T) versionableEntity.getEntity();

```

Comment:
versionableEntity might be null

Reply:
indeed, now using orElseThrow to raise a proper exception in such case.